apps/emqx/src/emqx_limiter/src/emqx_limiter_registry.erl

```diff
@@ -0,0 +1,208 @@
+            },
+            infinity
+        )
+    of
+        ok ->
+            ok;
+        {error, Reason} ->
+            error(Reason)
+    end.
+
+-spec unregister_group(group()) -> ok.
+unregister_group(Group) ->
+    gen_server:call(?MODULE, #unregister_group{group = Group}, infinity).
+
+-spec list_groups() -> [group()].
+list_groups() ->
+    gen_server:call(?MODULE, #list_groups{}, infinity).
+
+-spec find_group(group()) -> {module(), [{name(), emqx_limiter:options()}]} | undefined.
+find_group(Group) ->
+    case persistent_term:get(?PT_KEY(Group), undefined) of
+        undefined ->
+            undefined;
+        #group{module = Module, limiter_options = LimiterOptions} ->
+            {Module, maps:to_list(LimiterOptions)}
+    end.
+
+-spec get_limiter_options(limiter_id()) -> emqx_limiter:options() | no_return().
+get_limiter_options({Group, Name} = LimiterId) ->
+    case persistent_term:get(?PT_KEY(Group), undefined) of
+        #group{limiter_options = #{Name := LimiterOptions}} ->
+            LimiterOptions;
+        _ ->
+            error({limiter_not_found, LimiterId})
+    end.
+
+%%--------------------------------------------------------------------
+%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+init([]) ->
+    process_flag(trap_exit, true),
+    {ok, #{
+        group_names => sets:new([{version, 2}])
+    }}.
+
+handle_call(
+    #register_group{group = Group, module = Module, limiter_options = LimiterOptions} = Req,
+    _From,
+    #{group_names := GroupNames} = State
+) ->
+    OldGroup = persistent_term:get(?PT_KEY(Group), undefined),
+    case ensure_same_limiters(OldGroup, Req) of
+        ok ->
+            _ = persistent_term:put(?PT_KEY(Group), #group{
+                name = Group, module = Module, limiter_options = LimiterOptions
+            }),
+            {reply, ok, State#{
+                group_names := sets:add_element(Group, GroupNames)
+            }};
+        {error, _} = Error ->
+            {reply, Error, State}
+    end;
+handle_call(#unregister_group{group = Group}, _From, #{group_names := GroupNames} = State) ->
+    _ = persistent_term:erase(?PT_KEY(Group)),
+    {reply, ok, State#{
+        group_names := sets:del_element(Group, GroupNames)
+    }};
+handle_call(#list_groups{}, _From, #{group_names := GroupNames} = State) ->
+    {reply, sets:to_list(GroupNames), State};
+handle_call(Req, _From, State) ->
+    ?SLOG(error, #{msg => "unexpected_call", call => Req}),
+    {reply, ignore, State}.
+
+handle_cast(Req, State) ->
+    ?SLOG(error, #{msg => "unexpected_cast", cast => Req}),
+    {noreply, State}.
+
+handle_info(Req, State) ->
+    ?SLOG(error, #{msg => "unexpected_info", info => Req}),
+    {noreply, State}.
+
+terminate(_Reason, #{group_names := GroupNames} = _State) ->
+    ?SLOG(warning, #{msg => "terminating", group_names => sets:to_list(GroupNames)}),
```

Comment:
Maybe make this a debug level log, and also add a bit more information to `msg` to distinguish from whence it comes from?